File: app/src/main/java/com/mnj/java_solutions/FractionPairWithSumOne08082023.java
package com.mnj.java_solutions;

import java.util.HashMap;

public class FractionPairWithSumOne08082023 {

    public void countFractions() {
        int[] numerator = {3, 1, 12, 81, 2};
        int[] denominator = {9, 10, 18, 90, 5};
        int n = numerator.length;

        int count = 0;
        HashMap<String, Integer> map = new HashMap<>();

        for (int i = 0; i < n; i++) {
            int k = gcd(numerator[i], denominator[i]);
            int num = numerator[i] / k;
            int den = denominator[i] / k;

            String a = num + "/" + den;
            String d = den - num + "/" + den;

            if (map.containsKey(d)) {
                count = count + map.get(d);
            }
            map.put(a, map.getOrDefault(a, 0) + 1);
        }
        System.out.println("==>> number of pairs :" + count);
    }

    private int gcd(int a, int b) {
        if (a == 0)
            return b;
        return gcd(b % a, a);
    }
}

File: app/src/main/java/com/mnj/java_solutions/FractionPairWithSumOne08082023.kt
package com.mnj.java_solutions

class FractionPairWithSumOne08082023 {
    fun countFractions() {
        val numerator = intArrayOf(3, 1, 12, 81, 2)
        val denominator = intArrayOf(9, 10, 18, 90, 5)
        val n = numerator.size
        var count = 0
        val map = HashMap<String, Int>()
        for (i in 0 until n) {
            val k = gcd(numerator[i], denominator[i])
            val num = numerator[i] / k
            val den = denominator[i] / k
            val a = "$num/$den"
            val d = (den - num).toString() + "/" + den
            if (map.containsKey(d)) {
                count = count + map[d]!!
            }
            map[a] = map.getOrDefault(a, 0) + 1
        }
        println("==>> number of pairs :$count")
    }

    private fun gcd(a: Int, b: Int): Int {
        return if (a == 0) b else gcd(b % a, a)
    }
}
